validation: fix coins disappearing mid-package evaluation

glozow opened the pull request:
While we are evaluating a package, we split it into "subpackages" for evaluation (currently subpackages all have size 1 except the last one). If a subpackage has size 1, we may add a tx to mempool and call `LimitMempoolSize()`, which evicts transactions if the mempool gets full. We handle the case where the just-submitted transaction is evicted immediately, but we don't handle the case in which a transaction from a previous subpackage (either just submitted or already in mempool) is evicted. Mainly, since the coins created by the evicted transaction are cached in `m_view`, we don't realize the UTXO has disappeared until `CheckInputsFromMempoolAndCache` asserts that they exist. Also, the returned `PackageMempoolAcceptResult` reports that the transaction is in mempool even though it isn't anymore.

Fix this by not calling `LimitMempoolSize()` until the very end, and editing the results map with "mempool full" if things fall out.

Pointed out by instagibbs in https://github.com/bitcoin/bitcoin/commit/faeed687e5cde5e32750d93818dd1d4add837f24 on top of the v3 PR.

glozow committed:
[validation] add AcceptSubPackage to delegate Accept* calls and clean up m_view

(1) Call AcceptSingleTransaction when there is only 1 transaction in the
  subpackage. This avoids calling PackageMempoolChecks() which enforces
rules that don't need to be applied for a single transaction, i.e.
disabling CPFP carve out.
There is a slight change in the error type returned, as shown in the
txpackage_tests change.  When a transaction is the last one left in the
package and its fee is too low, this returns a PCKG_TX instead of
PCKG_POLICY. This interface is clearer; "package-fee-too-low" for 1
transaction would be a bit misleading.

(2) Clean up m_view and m_viewmempool so that coins created in this
sub-package evaluation are not available for other sub-package
evaluations. The contents of the mempool may change, so coins that are
available now might not be later.

glozow committed:
[validation] make PackageMempoolAcceptResult members mutable

After the PackageMempoolAcceptResult is returned from
AcceptMultipleTransactions, leave room for results to change due to
LimitMempool() eviction.

glozow committed:
[refactor] back-fill results in AcceptPackage

Instead of populating the last PackageMempoolAcceptResult with stuff
from results_final and individual_results_nonfinal, fill results_final
and create a PackageMempoolAcceptResult using that one.

A future commit will add LimitMempoolSize() which may change the status
of each of these transactions from "already in mempool" or "submitted to
mempool" to "no longer in mempool". We will change those transactions'
results here.

A future commit also gets rid of the last AcceptSubPackage outside of
the loop. It makes more sense to use results_final as the place where
all results end up.

glozow committed:
[validation] don't LimitMempoolSize in any subpackage submissions

Don't do any mempool evictions until package validation is done,
preventing the mempool minimum feerate from changing. Whether we submit
transactions separately or as a package depends on whether they meet the
mempool minimum feerate threshold, so it's best that the value not
change while we are evaluating a package.
This avoids a situation where we have a CPFP package in which
the parents meet the mempool minimum feerate and are submitted by
themselves, but they are evicted before we have submitted the child.

instagibbs commented:
@sipa you might want to weigh in as well

instagibbs commented on src/validation.cpp:
```diff
@@ -516,7 +516,7 @@ class MemPoolAccept
                             /* m_test_accept */ package_args.m_test_accept,
                             /* m_allow_replacement */ true,
-                            /* m_package_submission */ false,
+                            /* m_package_submission */ true, // do not LimitMempool
```
was this always supposed to be true?

glozow replied:
No, as it means skipping `LimitMempoolSize` within `Finalize()` and we're not always guaranteed to call `SubmitPackage` afterwards (e.g. if we quit early).

instagibbs reviewed (commented):
conditional ACK https://github.com/bitcoin/bitcoin/pull/28251/commits/c2c9dfe95b3c5b8332521d9a691725ee55ec3450

on having a test case catching the crash included. Tested with my own test case in https://github.com/bitcoin/bitcoin/pull/26403/commits/e3622f7ec944ec0e4c48ec3f29835d9f90d9d55c

ariard commented on src/validation.cpp:
```diff
@@ -554,6 +554,16 @@ class MemPoolAccept
+    /**
+     * Submission of a subpackage.  If only 1 transaction exists in subpackage, calls
+     * AcceptSingleTransaction() with adjusted ATMPArgs to avoid additional package policy
+     * restrictions like PackageMempoolChecks() and disabled RBF, and creates a
```
The “disabled RBF” can sound as confusing. There is no nsequence opted-out of the transaction candidate targeted by a replacement, it sounds rather to mean the `m_allow_replacement` of `ATMPArgs` which is true for `SingleInPackageAccept` and false for `PackageChildWithParents`.

glozow replied:
Edited comment

ariard commented on src/validation.cpp:
```diff
@@ -516,7 +516,7 @@ class MemPoolAccept
                             /* m_test_accept */ package_args.m_test_accept,
                             /* m_allow_replacement */ true,
-                            /* m_package_submission */ false,
+                            /* m_package_submission */ true, // do not LimitMempool
```
I think this could document more where the “do not LimitMempool” is applied, i.e in `Finalize()` L1123 which is itself called by `AcceptSingleTransaction` and the source of the bug. `LimitMempoolSize()` is called few times in validation.

glozow replied:
Added "in Finalize()" to the comment

ariard commented on src/validation.cpp:
```diff
@@ -1326,6 +1325,23 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+    ATMPArgs single_args = ATMPArgs::SingleInPackageAccept(args);
+    AssertLockHeld(::cs_main);
+    AssertLockHeld(m_pool.cs);
+    if (subpackage.size() > 1) {
```
I think this check doesn’t have test coverage as tx are individually validated in `AcceptPackage`.

ariard replied:
I think there is confusion, my comment is about the test coverage of the exact code path in the first half of `AcceptSubPackage` not of the second usage of `AccetSubPackage()`.

From a quick look on the tests at c2c9dfe9, it was uncertain.

ariard reviewed (commented):
Note the code path affected by the bug is under the `submitpackage` RPC, which is clearly indicated as an experimental RPC with an unstable interface.

More test coverage sounds good.

glozow committed:
[test framework] add ability to spend only confirmed utxos

Useful to ensure that the topologies of packages/transactions are as
expected, preventing bugs caused by having unexpected mempool ancestors.

glozow committed:
[refactor] split setup in mempool_limit test

We want to be able to re-use fill_mempool so that none of the tests
affect each other.

Change the logs from info to debug because they are otherwise repeated
many times in the test output.

glozow committed:
[test] mempool coins disappearing mid-package evaluation

Test for scenario(s) outlined in PR 28251.
Test what happens when a package transaction spends a mempool coin which
is fetched and then disappears mid-package evaluation due to eviction or
replacement.

fjahr reviewed (commented):
Concept ACK, will do a full review once the test has been added.

glozow commented:
Pushed a test - it should fail on master with the assertion.

Generally, the conditions to hit the case are:
- An almost-full mempool.
- `mempool_evicted_tx` (`parent`) A to-be-evicted tx that is submitted ahead of time. Has a very low feerate and is evicted immediately when mempool overflows. It has an output `coin_to_disappear`
- A package consisting of:
	- `cpfp_parent` (`child_a`) parent that spends `coin_to_disappear` and fails a check that requires its input to be loaded and makes it eligible for reconsideration (e.g. below mempool minimum feerate).
	- high-feerate parent (`child_b`), high feerate so that they are submitted individually and trigger eviction.
		- I've used 6 high-feerate parents in this test to make it easy to trigger overflow.
		- In the EA test, `child_b` triggers eviction due to proactive trimming of the sponsorless `parent` instead of mempool overflow.
	- `package_child` (`child_c`) spending the above parents.
		- The child must pay a fee high enough to CPFP the low-feerate parent. 
	- Additional requirements:
		- The low-feerate parent needs to be checked before the higher-feerate parents so that its inputs are loaded before the eviction happens. Otherwise it will just hit missing-inputs. This also means that as a side effect of the fee-based linearization in #26711 (which would put the low-feerate parent at the end), this test won't trigger. But I don't think the linearization is an adequate solution to this problem.
		- To make the test interesting, it should be low enough that the to-be-evicted tx is evicted regardless of when the eviction happens.

Timeline of what happens:
1. mempool is almost full
2. `mempool_evicted_tx` is submitted to mempool
3. `AcceptPackage(package)`:
	1. `AcceptSingleTransaction(cpfp_parent)`. This loads `coin_to_disappear` into `m_view` . After the fee is checked, the tx fails and we continue onto the next transactions.
	2. for each of the n high-feerate parents, `AcceptSingleTransaction(tx)`. These succeed and, in `Finalize()`, the transactions are submitted.
		1. (Without PR's changes) this also calls `LimitMempool()` which evicts by descendant score until the mempool is within size limits again.
		2. This kicks out `mempool_evicted_tx`, which means the `m_viewmempool` will no longer have a coin for `coin_to_disappear`. However, since the coin has already been cached in `m_view` (which has no backend), `m_view` doesn't know this.
	3. `AcceptSingleTransaction(package_child)` fails due to missing inputs.
	4. `AcceptMultipleTransactions([cpfp_parent, package_child]` 
		1. `PreChecks` looks for the inputs of each transaction, including the now-nonexistent `coin_to_disappear`. This succeeds because `m_view` has a coin for it.
		2. The CPFP works, so both transactions are submitted in `SubmitPackage()` , which calls `ConsensusScriptChecks()` which calls `CheckInputsFromMempoolAndCache()` which asserts that the coins all exist.
			1. We can't find the coin in the mempool (which is where it used to be).
			2. So we assume we will find it in the UTXO set. We call `AccessCoin()`  and assert `!coinFromUTXOSet.IsSpent()`. This fails because the coin does not exist (which is the same as isspent).

instagibbs commented on test/functional/mempool_limit.py:
```diff
@@ -84,13 +85,105 @@ def fill_mempool(self):
+        # This parent spends the above mempool transaction that exists when its inputs are first
+        # looked up, but disappears later. It is rejected for being too low fee (but eligible for
+        # reconsideration), and its inputs are cached. When the mempool transaction is evicted, its
+        # coin is no longer available, but the cache could still contains the tx.
```
```Suggestion
        # coin is no longer available, but the cache could still contain the tx.
```

instagibbs commented on test/functional/mempool_limit.py:
```diff
@@ -84,13 +85,105 @@ def fill_mempool(self):
+        assert_greater_than(mempoolmin_feerate, get_first_eviction_score(node))
+
+        package_hex = []
+        parent_utxos = []
```
```Suggestion
        parent_utxos = [] # utxos to be spent by the ultimate child transaction
```

instagibbs commented on test/functional/mempool_limit.py:
```diff
@@ -84,13 +85,105 @@ def fill_mempool(self):
+            # There is room for each of these transactions independently
+            assert node.testmempoolaccept([parent["hex"]])[0]["allowed"]
+
+        # Create a child spending everything, CPFPing the low-feerate parent just above mempool
```
```Suggestion
        # Create a child spending everything, CPFPing the low-feerate cpfp_parent just above mempool
```

instagibbs commented on test/functional/mempool_limit.py:
```diff
@@ -84,13 +85,105 @@ def fill_mempool(self):
 
+    def test_mid_package_eviction(self):
+        node = self.nodes[0]
+        self.log.info("Check a package where each parent passes the current mempoolminfee but would cause eviction")
```
```Suggestion
        self.log.info("Check a package where each parent passes the current mempoolminfee but would cause eviction before package submission terminates")
```

glozow replied:
added

instagibbs reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/28251/commits/3729ee483688da8427c9228de01407777d4e4691

The newly added test, while slightly obscure to trigger it, makes sense and triggers the crash bug when the fix is not applied.

![image](https://github.com/bitcoin/bitcoin/assets/5767891/cf6e6bb5-3b77-447e-8f16-6f21d93f95af)

glozow committed:
[CCoinsViewMemPool] track non-base coins and allow Reset

Temporary coins should not be available in separate subpackage submissions.
Any mempool coins that are cached in m_view should be removed whenever
mempool contents change, as they may be spent or no longer exist.

glozow commented on src/validation.cpp:
```diff
@@ -1184,32 +1184,21 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
 
-    // It may or may not be the case that all the transactions made it into the mempool. Regardless,
-    // make sure we haven't exceeded max mempool size.
-    LimitMempoolSize(m_pool, m_active_chainstate.CoinsTip());
```
Fixed, thanks

glozow commented on test/functional/mempool_limit.py:
```diff
@@ -84,13 +85,105 @@ def fill_mempool(self):
+        resulting_mempool_txids = node.getrawmempool()
+        assert mempool_evicted_tx["txid"] not in resulting_mempool_txids
+        assert cpfp_parent["txid"] not in resulting_mempool_txids
+        assert child["txid"] not in resulting_mempool_txids
```
Added this and the comment fixups

glozow commented:
Updated the approach + added a test for replacement case.

There are now 2 fixes:
(1) Clear any non-chainstate coins at the end of `AcceptSubPackage` so they can't be used in a different `AcceptSubPackage`.
(2) Don't trim in the middle of package evaluation.

Technically (1) is sufficient, but I think (2) is also beneficial as it avoids changing the mempool minimum feerate (which affects how we group transactions) mid-package evaluation. It would be sad if a CPFP package contains a parent that's just barely above mempool minimum feerate, so we submit it, then evict it (due to later parent), then can't submit the child... when we could have just submitted all of them and evicted something else.

instagibbs commented on src/txmempool.cpp:
```diff
@@ -982,6 +982,7 @@ bool CCoinsViewMemPool::GetCoin(const COutPoint &outpoint, Coin &coin) const {
     if (ptx) {
         if (outpoint.n < ptx->vout.size()) {
             coin = Coin(ptx->vout[outpoint.n], MEMPOOL_HEIGHT, false);
+            m_non_base_coins.emplace(outpoint);
```
this side-effect should be noted in the declaration comments

instagibbs commented on src/txmempool.h:
```diff
@@ -839,6 +839,12 @@ class CCoinsViewMemPool : public CCoinsViewBacked
     std::unordered_map<COutPoint, Coin, SaltedOutpointHasher> m_temp_added;
+
+    /**
+     * Set of all coins that have been fetched from mempool (not base). Used to internally track
```
```suggestion
     * Set of all coins that have been fetched from mempool or package (not base). Used to internally track
```

instagibbs reviewed (commented):
Ifirst pass on most recent fix to https://github.com/bitcoin/bitcoin/pull/28251/commits/a5a029c7799cb1f972e3fddc896bf009068cd8db

approach ack on fixes, would be flakey to trim things within packages prematurely.

instagibbs commented on src/validation.cpp:
```diff
@@ -1509,16 +1497,36 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
             Assume(it->second.m_result_type != MempoolAcceptResult::ResultType::INVALID);
             Assume(individual_results_nonfinal.count(wtxid) == 0);
+            if (!m_pool.exists(GenTxid::Txid(tx->GetHash()))) {
+                package_state_final.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
```
```Suggestion
                package_state_final.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
                results_final.emplace(wtxid, MempoolAcceptResult::Failure(mempool_full_state));
```
?

glozow committed:
[validation] return correct result when already-in-mempool tx gets evicted

Bug fix: a transaction may be in the mempool when package evaluation
begins (so it is added to results_final with MEMPOOL_ENTRY or
DIFFERENT_WITNESS), but get evicted due to another transaction
submission.

glozow commented on src/validation.cpp:
```diff
@@ -516,7 +516,7 @@ class MemPoolAccept
                             /* m_coins_to_uncache */ package_args.m_coins_to_uncache,
```
removed

instagibbs commented on src/validation.cpp:
```diff
@@ -1505,32 +1505,28 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
+            // Already-in-mempool transaction.
+            Assume(it->second.m_result_type != MempoolAcceptResult::ResultType::INVALID);
+            Assume(individual_results_nonfinal.count(wtxid) == 0);
+        } else if (const auto it{individual_results_nonfinal.find(wtxid)}; it != individual_results_nonfinal.end()) {
```
clearer the way it is now, thanks!

instagibbs reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/28251/commits/22fe13e3526b582170c05f823395dba0a3c9c99a

ariard commented on test/functional/mempool_limit.py:
```diff
@@ -34,51 +34,224 @@ def set_test_params(self):
+        assert_equal(node.getmempoolinfo()['minrelaytxfee'], Decimal('0.00001000'))
+        assert_equal(node.getmempoolinfo()['mempoolminfee'], Decimal('0.00001000'))
+
+        self.fill_mempool()
```
As I understand one of the condition to hit the case is a full mempool and when we call `LimitMempool()`, one interesting variation of the test case could be to limit the mempool size to the _subpackage_ only. It might be a tricky one, though boundary conditions are always nice to exercise.

glozow replied:
Yes, full mempool is one of the conditions (i.e. the one tested in `test_mid_package_eviction`). It's not strictly necessary though, the coin can also disappear due to replacement.

> one interesting variation of the test case could be to limit the mempool size to the subpackage only

Do you mean setting the mempool's maximum memory to something very tiny?

ariard reviewed (commented):
Thanks for the in-depth explanation of the conditions to hit the case. Will review more the code and test coverage as high-prio.

ariard commented on src/validation.cpp:
```diff
@@ -1382,21 +1429,12 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
-    // succeeds later (i.e. when evaluated with a fee-bumping child), the result changes (though not
-    // reflected in this map). If a transaction fails more than once, we want to return the first
-    // result, when it was considered on its own. So changes will only be from invalid -> valid.
+    // Stores "final" results from which we will create the returned PackageMempoolAcceptResult.
```
Finality is already an existent notion related to transaction validation (`IsFinalTx` in `src/consensus/tx_verify.cpp`). Maybe it can be called `results_nondefinitive` or `results_partial`.

glozow replied:
removed the word "final" and renamed this to `results_to_return`

glozow replied:
I agree renaming wouldn't be a bad idea, but perhaps the best time to do it is after #26711 (which changes the way this map is used) with a scripted-diff.

glozow commented on src/validation.cpp:
```diff
@@ -1326,6 +1325,54 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+    // If the transaction was submitted to the mempool, m_viewmempool will be able to fetch them from
+    // there. If it wasn't submitted to mempool, it is incorrect to keep them - future calls may try
+    // to spend those coins that don't actually exist.
+    // (2) Mempool coins also need to be removed. If the mempool contents have changed, coins
```
I mean the contents have changed because we submitted transaction(s), not due to another thread's actions.

glozow replied:
Clarified this comment

glozow commented on src/validation.cpp:
```diff
@@ -554,6 +554,17 @@ class MemPoolAccept
 
+    /**
+     * Submission of a subpackage. If only 1 transaction exists in subpackage, calls
+     * AcceptSingleTransaction() with adjusted ATMPArgs to avoid additional package policy
```
I think this is already the case? "calls AcceptSingleTransaction() with adjusted ATMPArgs"

glozow replied:
Resolving this comment but lmk if you meant something different

glozow commented:
> Concept ACK, will do a full review once the test has been added.

ping @fjahr :)

ariard commented:
If one wanna test the crash on master to check new test coverage, skipping the new fix branch here: https://github.com/ariard/bitcoin/commit/cf0b24ce0b7034ac6e344db7b04012f5b430d4a1

instagibbs commented:
Built a fuzzer which also caught this issue very quickly on my wimpy laptop: https://github.com/bitcoin/bitcoin/pull/28450

glozow commented on src/txmempool.h:
```diff
@@ -839,15 +839,27 @@ class CCoinsViewMemPool : public CCoinsViewBacked
     /** Add the coins created by this transaction. These coins are only temporarily stored in
      * m_temp_added and cannot be flushed to the back end. Only used for package validation. */
     void PackageAddTransaction(const CTransactionRef& tx);
+    /** Get all coins in m_non_base_coins and m_temp_added. */
```
Fixed comment

glozow commented on src/validation.cpp:
```diff
@@ -1174,32 +1184,21 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
-    // Find the wtxids of the transactions that made it into the mempool. Allow partial submission,
-    // but don't report success unless they all made it into the mempool.
+
+    // Add successful results. The returned results may change later if LimitMempool() evicts them.
```
Thanks, changed

glozow commented on src/validation.cpp:
```diff
@@ -639,10 +650,9 @@ class MemPoolAccept
     // cache - should only be called after successful validation of all transactions in the package.
-    // The package may end up partially-submitted after size limiting; returns true if all
-    // transactions are successfully added to the mempool, false otherwise.
+    // Does not call LimitMempool(), so maximum size may be temporarily exceeded.
```
mentioned

glozow reviewed (commented):
Incorporated suggestions. All documentation-related so should be trivial to re-ack.

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/28251/commits/a929cab981688dae88f1ae692a6a322328da12b0

trivial comment updates

glozow commented:
Thanks! Reproduced, debugging

glozow committed:
[policy] check for duplicate txids in package

Duplicates of normal transactions would be found by looking for
conflicting inputs, but this doesn't catch identical empty transactions.
These wouldn't be valid but exiting early is good and AcceptPackage's
result sanity checks assume non-duplicate transactions.

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/28251/commits/7d7f7a1189432b1b6245ba25df572229870567cb

inclusion of duplicate-txid check in `CheckPackage` which prevents the assertion being hit for the 2nd of the duplicate txids

sdaftuar commented:
I'm reviewing this PR more carefully, after seeing @instagibbs' comment here: https://github.com/bitcoin/bitcoin/pull/31122#discussion_r1813272637

As I understand it, I think there are two key behaviors introduced in this PR:
1) Clearing state between subpackage evaluations (specifically, clearing the coins in `m_viewmempool`)
2) Deferring mempool limiting until the end of package validation

However, the specific issue that @glozow described above (https://github.com/bitcoin/bitcoin/pull/28251#issuecomment-1688170869), where we need to ensure that package validation **fails** if an input is evicted mid-validation, seems to be completely mitigated by behavior (1).  And this behavior is tested in the `mempool_limit` test introduced in this PR.

As far as I can tell from reviewing [`3ea71fe` (#28251)](https://github.com/bitcoin/bitcoin/pull/28251/commits/3ea71feb11c261f002ed918f91f3434fd8a23589), behavior (2) is designed to allow package validation to **succeed** in a case where we have, say, a child `C` with two parents `A` and `B`, and we submit a package `[A, B, C]` and potentially trigger this outcome:
* `A` is accepted singly
* `B` is accepted singly, but the mempool overflows and `A` is trimmed
* `C` is rejected for missing inputs

So by preventing mempool limiting from taking place mid-package, the choice of what to evict from the mempool is deferred until after `C` is able to make it in, potentially changing the lowest-descendant-score transaction in the mempool from `A` to something else (as `A`'s descendant score might be pretty high after `C` is accepted).

However, I don't think this PR introduced any test for this second behavior, of ensuring success is possible even if an earlier parent might be trimmed mid-validation.

Does this summary sound right to others? Have I overlooked other behavior changes/bugfixes that were introduced here?

glozow commented:
> As I understand it, I think there are two key behaviors introduced in this PR

> However, the specific issue that @glozow described above (https://github.com/bitcoin/bitcoin/pull/28251#issuecomment-1688170869), where we need to ensure that package validation fails if an input is evicted mid-validation, seems to be completely mitigated by behavior (1).

> As far as I can tell from reviewing [3ea71fe (#28251)](https://github.com/bitcoin/bitcoin/pull/28251/commits/3ea71feb11c261f002ed918f91f3434fd8a23589), behavior (2) is designed to allow package validation to succeed in a case where...

> However, I don't think this PR introduced any test for this second behavior,

This looks accurate to me. I've looked a little at the tests (it's been a while :sweat_smile:) and can't really point to one for deferring evictions. Will review #31152.
